utils/results/common.go

```diff
@@ -53,18 +53,110 @@ func NewFailBuildError() error {
 	return coreutils.CliError{ExitCode: coreutils.ExitCodeVulnerableBuild, ErrorMsg: "One or more of the detected violations are configured to fail the build that including them"}
 }
 
-// In case one (or more) of the violations contains the field FailBuild set to true, CliError with exit code 3 will be returned.
-func CheckIfFailBuild(results []services.ScanResponse) bool {
-	for _, result := range results {
-		for _, violation := range result.Violations {
-			if violation.FailBuild {
+// This func iterates every violation and checks if there is a violation that should fail the build.
+// The build should be failed if there exists at least one violation in any target that holds the following conditions:
+// 1) The violation is set to fail the build by FailBuild or FailPr
+// 2) The violation has applicability status other than 'NotApplicable' OR the violation has 'NotApplicable' status and is not set to skip-non-applicable
+func CheckIfFailBuild(auditResults *SecurityCommandResults) (bool, error) {
+	for _, target := range auditResults.Targets {
+		// We first check if JasResults exist so we can extract CA results and consider Applicability status when checking if the build should fail.
+		if target.JasResults == nil {
+			shouldFailBuild := checkIfFailBuildWithoutConsideringApplicability(target)
+			if shouldFailBuild {
+				// If we found a violation that should fail the build, we return true.
+				return true, nil
+			}
+		} else {
+			// If JasResults are not empty we check old and new violation while considering Applicability status and Skip-not-applicable policy rule.
+			var shouldFailBuild bool
+			if err := checkIfFailBuildConsideringApplicability(target, auditResults.EntitledForJas, &shouldFailBuild); err != nil {
+				return false, fmt.Errorf("failed to check if build should fail for target %s: %w", target.ScanTarget.Target, err)
+			}
+			if shouldFailBuild {
+				// If we found a violation that should fail the build, we return true.
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func checkIfFailBuildConsideringApplicability(target *TargetResults, entitledForJas bool, shouldFailBuild *bool) error {
```

Comment:
I think you can combine the logic and have the `checkIfShouldFailBuildAccordingToPolicy` have the logic of the 'if not applicable'